[ML] New change point detection

tveasey opened the pull request:
This reworks change point detection to makes use of the enhancements to `CTimeSeriesSegmentation` introduced in #1543 and #1544. It also revisits and ensures that all disabled tests are enabled and passing.

This is a non-issue since it will not be documented separately.

~(This is WIP since I expect to have to make some more tweaks to `CTimeSeriesTestForChange` and I have haven't yet fixed and re-enabled all unit tests.)~

droberts195 commented on lib/maths/CTimeSeriesDecompositionDetail.cc:
```diff
@@ -292,7 +292,7 @@ const core::TPersistenceTag MEAN_OFFSET_7_11_TAG{"c", "mean_offset"};
 const core::TPersistenceTag LARGE_ERROR_FRACTION_7_11_TAG{"e", "large_error_fraction"};
 const core::TPersistenceTag LAST_TEST_TIME_7_11_TAG{"f", "last_test_time"};
-const core::TPersistenceTag LAST_CHANGE_TIME_7_11_TAG{"g", "last_change_time"};
+const core::TPersistenceTag LAST_UNCOMMITTED_CHANGE_TIME_7_11_TAG{"g", "last_uncommitted_change_time"};
```
I know it's not directly related to this change, but if you haven't done so already please can you update `api::CAnomalyJob::MODEL_SNAPSHOT_MIN_VERSION` to `7.11.0` on your feature branch so we won't try to load the new format on old nodes.

tveasey replied:
Good point: it's on my list. I was going to bump it at the point I merged to master in case this slipped, but I think that's unlikely now so I can add in this PR.

tveasey replied:
Done in [3617328](https://github.com/elastic/ml-cpp/pull/1550/commits/36173280d512decd1340e2e2e3b7d25771efeabf).

valeriy42 commented on lib/maths/CTimeSeriesSegmentation.cc:
```diff
@@ -665,21 +673,20 @@ void CTimeSeriesSegmentation::fitPiecewiseLinearScaledSeasonal(
         for (std::size_t j = 1; j < segmentation.size(); ++j) {
-            TMeanAccumulator projection;
-            TMeanAccumulator Z;
+            TMeanAccumulator xp;
+            TMeanAccumulator pp;
```
Do the new variable names `xp` and `pp` refer to some formula? For me personally, it was easier to understand what `projection` and `Z` meant.

tveasey replied:
It was just to make these consistent with other code which does a similar thing, i.e. I went for consistency. I'm happy to instead standardise on these names.

tveasey replied:
Reverted in [079a076](https://github.com/elastic/ml-cpp/pull/1550/commits/079a07614ed72f3b9562f478b634eaedbcd68f77).

valeriy42 commented on lib/maths/CTimeSeriesTestForSeasonality.cc:
```diff
@@ -374,6 +376,7 @@ CSeasonalDecomposition CTimeSeriesTestForSeasonality::decompose() const {
                     trend.add(static_cast<double>(j), CBasicStatistics::mean(values[j]),
                               CBasicStatistics::count(values[j]));
                 }
+                trend.parameters(parameters);
```
I am probably mistaken, but it seems to me that `parameters` is getting set but not read from here on. Why do we need to store it if we don't use the value?

tveasey replied:
This is correct, but the parameters are referenced by the predictor. This is actually a significant performance optimisation since the runtime is dominated by computing these parameters, I'll add a comment to explain what is going on here.

tveasey replied:
See comment in [079a076](https://github.com/elastic/ml-cpp/pull/1550/commits/079a07614ed72f3b9562f478b634eaedbcd68f77).

valeriy42 commented on lib/maths/unittest/CTimeSeriesModelTest.cc:
```diff
@@ -30,6 +30,7 @@
 #include <test/CTimeSeriesTestData.h>
 
 #include "TestUtils.h"
+#include "maths/MathsTypes.h"
```
```suggestion
#include <TestUtils.h>
#include <maths/MathsTypes.h>
```

tveasey replied:
See [079a076](https://github.com/elastic/ml-cpp/pull/1550/commits/079a07614ed72f3b9562f478b634eaedbcd68f77).

valeriy42 reviewed (commented):
Looks good altogether :+1: I am wondering if as a part of this change in anomaly detection, we could move the relevant files (e.g. `CTimeSeries*`) into a subdirectory to have more structure?

tveasey commented on lib/maths/unittest/CTimeSeriesTestForChangeTest.cc:
```diff
@@ -0,0 +1,242 @@
+    TMeanAccumulator meanError;
+    TMeanAccumulator meanTimeError;
+
+    TDoubleVec samples;
```
The random number generator calls the reserve for us when we actually fill in the values.

tveasey commented on include/maths/CTimeSeriesDecompositionDetail.h:
```diff
@@ -85,10 +92,13 @@ class MATHS_EXPORT CTimeSeriesDecompositionDetail : private CTimeSeriesDecomposi
-        //! The window values preconditioner to use for seasonality testing.
-        TFilteredPredictor s_SeasonalityTestPreconditioner;
+        //! The time series decomposition.
+        CTimeSeriesDecomposition* s_Decomposition;
```
This is a temporary object and there is no ownership involved, it just needs a reference back to the decomposition. The constructor takes by reference but it is stored as a pointer so this type is copyable.

tveasey commented on include/maths/CTimeSeriesDecompositionDetail.h:
```diff
@@ -181,8 +204,111 @@ class MATHS_EXPORT CTimeSeriesDecompositionDetail : private CTimeSeriesDecomposi
-    //! \brief Scans through increasingly low frequencies looking for custom
-    //! diurnal and any other large amplitude seasonal components.
+    //! \brief Checks for sudden change or change events.
+    class MATHS_EXPORT CChangePointTest : public CHandler {
```
For definite this needs restructuring. I want to create subdirectories in `maths` as you suggest. Then this can all move into a details subdirectory, for time series modelling, and be split over several files. However, I don't want to make this change in this PR, which is already large. It also needs us to do some work to get the build system to accept the new directory arrangement. So whilst I agree with this comment, I would propose making the change in a separate PR. WDYT?

valeriy42 replied:
+1 Let's have it in a separate PR

tveasey commented on lib/maths/unittest/CTimeSeriesTestForChangeTest.cc:
```diff
@@ -0,0 +1,242 @@
+#include <test/CRandomNumbers.h>
+
+#include "TestUtils.h"
+#include "core/CContainerPrinter.h"
```
Local files to test directory are normally included like this, but I'll correct `CContainerPrinter`.

tveasey replied:
See [079a076](https://github.com/elastic/ml-cpp/pull/1550/commits/079a07614ed72f3b9562f478b634eaedbcd68f77).

tveasey commented:
>  we could move the relevant files (e.g. CTimeSeries*) into a subdirectory to have more structure?

Definitely, however I would prefer to make this in a separate PR.

tveasey commented on lib/maths/unittest/CTimeSeriesTestForChangeTest.cc:
```diff
@@ -0,0 +1,242 @@
+    TMeanAccumulator meanTimeError;
+
+    TDoubleVec samples;
+    for (std::size_t test = 0; test < 100; ++test) {
```
Done in [079a076](https://github.com/elastic/ml-cpp/pull/1550/commits/079a07614ed72f3b9562f478b634eaedbcd68f77).

tveasey commented:
Testing has shown up a couple of refinements I need to make. I'm going to prepare one more change which addresses these issues. I'll let you know as soon as its ready for review again.

valeriy42 reviewed (approved):
LGTM

tveasey commented:
This passes CI on windows and linux and we'd like to kick off another test over so I'm going to go ahead and merge.
